// @codemod/core/src/PrettierPrinterPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resolve_1 = require("resolve");
const RecastPlugin_1 = require("./RecastPlugin");
function loadPrettier() {
    try {
        return require(resolve_1.sync('prettier', { basedir: process.cwd() }));
    }
    catch (_a) {
        return require('prettier');
    }
}
const DEFAULT_PARSER = 'babel';
function default_1() {
    const prettier = loadPrettier();
    function inferParser(filepath) {
        if (!filepath) {
            return DEFAULT_PARSER;
        }
        const { inferredParser } = prettier.getFileInfo.sync(filepath);
        return (inferredParser || DEFAULT_PARSER);
    }
    function resolvePrettierConfig(filepath) {
        return Object.assign(Object.assign({ parser: inferParser(filepath) }, (typeof filepath === 'string'
            ? prettier.resolveConfig.sync(filepath)
            : undefined)), { filepath });
    }
    return {
        parserOverride: RecastPlugin_1.parse,
        generatorOverride(ast, options) {
            return {
                code: prettier.format(RecastPlugin_1.generate(ast).code, resolvePrettierConfig(options.filename)),
            };
        },
    };
}
exports.default = default_1;
